### PR TITLE
New component to embed a view controller in a card-like view

### PR DESCRIPTION
**Pull request by eshurakov**
### Details

This PR add a new `SPCardViewController` component, which can embed another view controller in a view with rounded corners, shadow and a little bit transparent background.
This component will be used to wrap and show new history slider on the note editor screen, but not used anywhere at the moment.

### Notes
This is part of https://github.com/Automattic/simplenote-ios/issues/762
Will follow up with another PR for History Slider

### Screenshots

These are the example of how the component looks like.

<p float="left">

</p>

### CODE REVIEW

File: Simplenote/Classes/SPCardViewController.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    func addShadowView() {
+        let shadowView = SPShadowView(cornerRadius: Constants.cornerRadius,
+                                      roundedCorners: [.topLeft, .topRight])
+        view.addFillingSubview(shadowView)
+    }
+
+    func setupContainerView() {
+        view.addFillingSubview(containerView)
+
+        containerView.backgroundColor = UIColor.simplenoteBackgroundColor.withAlphaComponent(0.97)
```
Comment (jleandroperez) [inline Simplenote/Classes/SPCardViewController.swift:42]: Nipticky: Maybe `0.97` > constant?

Comment (jleandroperez) [reply Simplenote/Classes/SPCardViewController.swift:42]: Detail: You may need to listen to `VSThemeManagerThemeDidChangeNotification` and refresh the `.backgroundColor`.

TL;DR: Simplenote has supported, for a few years, dark mode. As per iOS 13 you don't need to do anything, for legacy support would require a manual refresh.

File: Simplenote/Classes/SPCardViewController.swift [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        containerView.layer.cornerRadius = Constants.cornerRadius
+        containerView.layer.maskedCorners = [.layerMaxXMinYCorner, .layerMinXMinYCorner]
+        containerView.layer.masksToBounds = true
+    }
+
+    func addChildViewController() {
+        addChild(viewController)
+        containerView.addFillingSubview(viewController.view)
```
Comment (jleandroperez) [inline Simplenote/Classes/SPCardViewController.swift:50]: Out of curiosity, do you think it's possible to avoid using `containerView` entirely, and add the `viewController.view` as a direct subview?

File: Simplenote/UIView+Constraints.swift [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+extension UIView {
+    func addFillingSubview(_ view: UIView,
+                           edgeInsets: UIEdgeInsets = .zero,
+                           useSafeArea: Bool = false) {
+
+        view.translatesAutoresizingMaskIntoConstraints = false
+
+        addSubview(view)
+        view.frame = bounds.inset(by: edgeInsets)
```
Comment (jleandroperez) [inline Simplenote/UIView+Constraints.swift:11]: Since the insets are set via constraints, maybe we can get rid of this one?

File: Simplenote/SPShadowView.swift [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+    override var bounds: CGRect {
+        didSet {
+            updatePath()
+        }
+    }
+
+    /// Designated Initializer
+    ///
+    ///  - cornerRadius: The radius of each corner oval
```
Comment (jleandroperez) [inline Simplenote/SPShadowView.swift:20]: Nipticky: If you add `// - Parameters:` before enumerating the method's parameters, Xcode's **Option + Click** will display them under the title **Parameters**

File: Simplenote/SPShadowView.swift [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        configureShadow()
+        configureMask()
+
+        updatePath()
+    }
+
+    /// Configuration of the shadow
+    ///
+    func configureShadow() {
+        layer.shadowColor = UIColor.black.cgColor
```
Comment (jleandroperez) [inline Simplenote/SPShadowView.swift:55]: Maybe it'd be useful to parametrize this one? 🤔 

(In the lines of `var shadowColor = UIColor.black`?)

File: Simplenote/SPShadowView.swift [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+}
+
+// MARK: - Constants
+//
+private extension SPShadowView {
+    struct Constants {
+        static let shadowOpacity: Float = 0.1
+        static let shadowRadius: CGFloat = 4.0
+        static let shadowOffset = CGSize(width: 0, height: -2)
```
Comment (jleandroperez) [inline Simplenote/SPShadowView.swift:95]: You can use UIOffset, instead, which might be more appropriate, in this particular case!

File: Simplenote/Classes/SPCardViewController.swift [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import UIKit
+
+// MARK: - SPCardViewController
```
Comment (jleandroperez) [inline Simplenote/Classes/SPCardViewController.swift:3]: Nipticky: Mind adding a brief description of what's the purpose of this VC?

File: Simplenote/Classes/SPCardViewController.swift (hunk 1 above)
Comment (eshurakov) [reply Simplenote/Classes/SPCardViewController.swift:42]: Thanks for the info about `VSThemeManagerThemeDidChangeNotification `. I wasn't aware of it, will add support for it!

File: Simplenote/Classes/SPCardViewController.swift (hunk 2 above)
Comment (eshurakov) [reply Simplenote/Classes/SPCardViewController.swift:50]: I believe it's not possible. Right now card consists of 2 subviews: 
* shadow view, which draws shadow on the outside of bounds rect
* content view, which has rounded corners and clips inside content, so that child view controller doesn't need to worry about any of those

If we would add child view controller directly to the main view, we would need to apply rounded corners and clip subviews. This would clip and remove shadow.

File: Simplenote/SPShadowView.swift (hunk 4 above)
Comment (eshurakov) [reply Simplenote/SPShadowView.swift:20]: Nice! Didn't know about that, thanks!

File: Simplenote/SPShadowView.swift (hunk 5 above)
Comment (eshurakov) [reply Simplenote/SPShadowView.swift:55]: Good point

File: Simplenote/SPShadowView.swift (hunk 6 above)
Comment (eshurakov) [reply Simplenote/SPShadowView.swift:95]: I agree UIOffset would fit better there, but it is used to configure `layer.shadowOffset`, which is `CGSize` :-(

File: Simplenote/Classes/SPCardViewController.swift (hunk 7 above)
Comment (eshurakov) [reply Simplenote/Classes/SPCardViewController.swift:3]: Good point! Will add a description 👍

File: Simplenote/UIView+Constraints.swift (hunk 3 above)
Comment (eshurakov) [reply Simplenote/UIView+Constraints.swift:11]: Indeed 👍

**Comment by eshurakov** [comment]
Thank you for the feedback, @jleandroperez. I've addressed some of the comments. Please take a look when you have time.

**Comment by jleandroperez** [comment]
Merging this one, thank you @eshurakov !!
